### ISSUE TITLE
Tracking issue for `late_bound_lifetime_arguments` compatibility lint

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#42868]
cc https://github.com/rust-lang/rust/pull/42492

TODO: write the description

#### How to fix this warning/error

Just removing the lifetime arguments pointed to by the lint should be enough in most cases.

#### Current status
- [x] https://github.com/rust-lang/rust/pull/42492 introduces the `late_bound_lifetime_arguments` lint as warn-by-default
- [ ] PR ? makes the `late_bound_lifetime_arguments` lint deny-by-default
- [ ] PR ? makes the `late_bound_lifetime_arguments` lint a hard error

**Comment by SimonSapin** [comment]
```rust
error: cannot specify lifetime arguments explicitly if late bound lifetime parameters are present
   --> /home/travis/build/servo/servo-with-rust-nightly/servo/components/style/stylesheets/stylesheet.rs:243:27
    |
243 |         self.iter_rules::<'a, 'b, EffectiveRules>(device, guard)
    |                           ^^
    |
note: lint level defined here
   --> /home/travis/build/servo/servo-with-rust-nightly/servo/components/style/lib.rs:26:9
    |
26  | #![deny(warnings)]
    |         ^^^^^^^^
    = note: #[deny(late_bound_lifetime_arguments)] implied by #[deny(warnings)]
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #42868 <https://github.com/rust-lang/rust/issues/42868>
```

What does "late bound" mean in this context? What should we do instead?

**Comment by petrochenkov** [comment]
>What does "late bound" mean in this context? 

Well, that's a good question, but not something that can be explained in an error message.
See for example http://smallcultfollowing.com/babysteps/blog/2013/10/29/intermingled-parameter-lists/#early--vs-late-bound-lifetimes. I believe nomicon had something about it, but I can't find it right now.

This issue should have a description, and @nikomatsakis is probably the best person to write it, but I can try as well.

>What should we do instead?

Remove `'a, 'b, `, it never did anything anyway.

**Comment by petrochenkov** [comment]
The error and lint should probably have an extra note pointing to one of these late bound lifetime parameters.

**Comment by SimonSapin** [comment]
>> What should we do instead?
>
> Remove `'a, 'b`, it never did anything anyway.

That’s not true in the case of https://github.com/rust-lang/rust/issues/42508, where it is necessary to make something compile at all.

**Comment by petrochenkov** [comment]
One possible solution for the problem of intermixed late- and early-bound arguments is to support explicit `for<...>` clauses on function items in the same way it's done on function pointers:

```rust
for<'late> fn f<'early1, 'early2>(a: &'late u8, b: &'early1 u8, c: &/*'elided*/u8) -> &'early2 u8 { ... }
```
If `for` is specified then all lifetimes defined in it are considered late-bound, and all lifetimes specified in generic parameters are considered early-bound.
For function pointers the equivalent construction is already supported:
```rust
type F<'early1, 'early2> = for<'late> fn(&'late u8, &'early1 u8, &/*'elided*/u8) -> &'early2 u8;
```

The function could then be called with explicitly specified lifetime arguments, despite the presence of late-bound parameters:
```rust
f::<'arg_early1, 'arg_early2>()
```

**Comment by nikomatsakis** [comment]
@SimonSapin 

I believe there exists a workaround to force early-bound lifetimes at a time when they are appropriate. Are you blocked on this point? In particular, you could do something like `where 'static: 'a`, which would force `'a` to be early bound (because it appears in a where-clause). 

The definition of late bound lifetimes, really, is that they are those lifetimes that both:

- Lifetimes that appear in argument types
- and do **not** appear in where clauses.

All other lifetimes bound on a function are early-bound. So if you have a dummy where clause like `where 'static: 'a`, which has no particular meaning, it does force `'a` to be early-bound.

I do hope we can clean this up in the future -- perhaps in another epoch -- but for now that is how it works. There are some annoying interactions with inference that make this all hard to alter in a totally backwards compatible way.

**Comment by SimonSapin** [comment]
I am stuck in the sense that I have warnings that I don’t know how to work around, don’t understand in the first place, and that say “it will become a hard error in a future release”.

**Comment by arielb1** [comment]
I think you could use something like `(parse_nested_block: fn(&mut Parser<'i, 't>, _))(...)` (using type ascription) instead of passing generic parameters. Ugly, but works.

**Comment by arielb1** [comment]
Nope, that doesn't work for some reason.

**Comment by pnkfelix** [comment]
@petrochenkov wrote (in the description):

> It's not clear how to provide arguments for early-bound lifetime parameters if they are intermixed with late-bound parameters in the same list. For now providing any explicit arguments is prohibited if late-bound parameters are present, so in the future we can adopt any solution without hitting backward compatibility issues.
> 
> Note that late-bound lifetime parameters can be introduced implicitly through lifetime elision
> \[...\]

While each of these paragraphs made sense to me when I read them on my own, I am curious about the conclusion that it leads to.

Consider a case like this:

```rust
fn all_explicits_are_early<'a, 'b>(arg: &Foo<'a, 'b>) -> Bar<'a, 'b> where 'a: 'a, 'b: 'b { ... }
```

I would think that a call expression `all_expilcits_are_early::<'x, 'y>(param)` would be sensible: we are providing arguments for the explicit parameters, which are all early bound.

But unfortunately, the above function definition actually desugars into something with a hidden late-bound lifetime parameter, due to lifetime elision on the `arg: &Foo<'a, 'b>`. Which then causes the error here to fire in the case described above.

 * (This ends up being one part of why some workarounds to this are not applying to @SimonSapin's example from 42508 )

Would it be possible to loosen the error checking here so that if all the explicitly listed lifetime parameters are early-bound, then one still gets to specify the lifetimes explicitly at the call-site, regardless of any hidden late bound lifetimes introduced due to the elision rules?

**Comment by nikomatsakis** [comment]
@pnkfelix 

> Would it be possible to loosen the error checking here so that if all the explicitly listed lifetime parameters are early-bound, then one still gets to specify the lifetimes explicitly at the call-site, regardless of any hidden late bound lifetimes introduced due to the elision rules?

That would be possible, I think. It would commit us to supporting that interpretation, of course.

**Comment by petrochenkov** [comment]
@pnkfelix
The logic (beside just being the most conservative variant) is that elided lifetimes are purely a sugar for the explicitly written form, i.e.
```rust
// Full form
fn foo<'early: 'early, 'late>(arg: &'late u8) { ... }
// Short form
fn foo<'early: 'early>(arg: &u8) { ... }
```
are equivalent and can be switched freely without changing any observable behavior for `foo`'s users.

If the short form doesn't report an error and the full form does, then it's not a sugar and you can't switch between two forms backward-compatibly.

**Comment by petrochenkov** [comment]
With my proposed solution (https://github.com/rust-lang/rust/issues/42868#issuecomment-325647940) elided lifetimes can always be written explicitly without affecting observable behavior for users, so the restriction is unnecessary and can be relaxed

```rust
// Short form
fn foo<'early: 'early>(arg: &u8) { ... } // `foo::<'static>` is okay
// Full desugared form
for<'late> fn foo<'early: 'early>(arg: &'late u8) { ... } // `foo::<'static>` is still okay
```

**Comment by SimonSapin** [comment]
This came up again. https://crates.io/crates/cssparser still emits this warning, and I still don’t know of a way to fix it without getting borrowck errors instead.

**Comment by cole-miller** [comment]
It seems this shows up as either a warn-level lint or a hard error depending on the code.

Example that triggers a warning (on stable and nightly):

```rust
static X: u8 = 0;
fn f<'a, 'b>(_x: &'b u8) -> &'a u8  { &X }
fn main() { let _ = f::<'static>; }
```

Example that triggers a hard error (what led me to open #80618; again, stable or nightly):

```rust
fn f<'a>() {}
fn main() { let _ = f::<'static>; }
```

Is this expected?

**Comment by petrochenkov** [comment]
@cole-miller 
It's reported as a lint only in cases where making it a hard error would cause breakage back in 2017.
The intent is to produce an error, in general, the lint is only a way to give people some time to fix code.

**Comment by cole-miller** [comment]
Thanks, understood. So I can go ahead and assign an error number and `--explain` text?

**Comment by fishrockz** [comment]
For anyone else finding things like this you might find this helpful

It was actually solved at the call sight rather than the function

``` diff
-lister::<'a, T>(buffer, element.clone(), project);
+lister::<T>(buffer, element.clone(), project);
```

Normally rustc gives the most excellent tips but i think in this case we could do better..

**Comment by RalfJung** [comment]
I have to say I am quite confused by the issue description here, and I don't understand what the underlying issue is. The core of my confusion seems to be captured by this (repeated) statement

> the first argument doesn't make much sense because the function is not parameterized by 'a.

I strongly disagree! The function *is* parameterized by `'a`. Sure, this parameter is a type system fiction (the actual runtime address of the function is the same for all choices of `'a`), but that doesn't make `'a` any less of a parameter. It can still be valuable to explicitly set that parameter to guide inference.

Taking a step back: The fact that some lifetime parameters are type system fictions while others are actually relevant for monomorphization is very confusing and surprising. (I have never fully understood why this is the case; I'll need to dig into this some time.) This will clearly incur some friction when creating function pointers since monomorphization-relevant ("early-bound") lifetime parameters need to be chosen the moment the function pointer is created. I can see that we might want to improve that situation. But reading just this issue I really don't understand the motivation for the current change, since there's a big gap between "unfortunately some lifetime parameters actually matter for monomorphization" and "other lifetime parameters (that don't matter for monomorphization) don't make sense and we should disallow people from explicitly stating them". Which problem is being solved here?

**Comment by amanlai** [comment]
A related [Stack Overflow discussion](https://stackoverflow.com/q/66302521/19123103).
